### ISSUE TITLE
RFC Vendor shared libraries to the wheel by default

### ISSUE THREAD

**Issue by ryanking13** [pyodide/pyodide#4841]
## Background

Many Python packages rely on dynamic libraries at runtime. For example, scipy uses openBLAS, and pyproj dynamically links to libproj. Managing these libraries can be challenging because they are not Python packages and cannot be uploaded to PyPI, although they are necessary dependencies from a Python package perspective.

In the Python packaging world, some approaches have been proposed to address this, including allowing native libraries outside of Python packages to be uploaded (e.g., conda) and embedding these libraries inside the package (e.g., the [manylinux](https://peps.python.org/pep-0513/) PEP).

Currently, I believe the manylinux proposal is working well, allowing most packages to be easily installed with pip.

## Current Status in Pyodide

Pyodide currently employs a mix of both approaches. Pyodide recipes support building libraries alongside Python packages, and pyodide-lock.json defines dependencies between these Python packages and libraries, enabling them to be loaded simultaneously.

Additionally, Pyodide recipes support embedding libraries that depend on Python packages inside the wheel file, similar to manylinux wheels. This is achieved by enabling the `vendor-sharedlib` option in the meta.yaml file:

```yaml
build:
  vendor-sharedlib: true
```

when this flag is enabled, the [`auditwheel-emscripten` library](https://github.com/ryanking13/auditwheel-emscripten) analyzes the dependencies of the Python package and inserts the library into the wheel file.

## Problems with vendoring shared libs

For now, the `vendor-sharedlib` option is not enabled by default because vendoring libraries into a wheel file can lead to duplicate copies of the same library across different packages. This can increase the overall package size, especially if multiple packages use the same library. For instance, if both scipy and sparseqr (which depends on scipy and openblas) use the vendor-sharedlib option, openblas would be included in both packages. Then, users who use `sparseqr` would download the openblas twice, resulting in a slow package installation time.

In a native environment, this is not a significant issue because the package only needs to be downloaded once, and disk space is not a major concern (nobody will care if they consume 10 more megabytes). However, in a browser environment, packages need to be downloaded each time, and reducing the overall package size is crucial (at least, we thought it was crucial). Thus, avoiding duplicate libraries in different packages has been a valid approach for Pyodide.

## The situation is changing

I think it's time to reconsider this decision. We are increasingly moving towards out-of-tree builds and encouraging individual package maintainers to build their own packages (thanks to @agriyakhetarpal).

In the long run, if Pyodide packages are uploaded to PyPI or other Python package indexes, users will want the packages they install from these indexes to work seamlessly. The current strategy of deploying libraries separately from the package, which is incompatible with the package index, would no longer be viable. This would make it difficult for users to install the package (as Python developers probably did before the introduction of wheel and manylinux).

Therefore, I propose changing the default setting of pyodide-build to vendor the library in the package. Even if this results in duplicate libraries being included in the package.

## What Needs to Be Done

We'll need to change the following things:

1. Change the default option to include the library in the package when pyodide-build builds a recipe.

2. Enhance library support in out-of-tree builds (`pyodide build`). Currently, linking external libraries and vendoring them into the package during an out-of-tree build is challenging and needs improvement.

3. (Optional) In `pyodide-lock`, we may consider extracting the library inside the package and making it a separate shared library: "unvendoring" the vendored libraries, which could optimize the size of packages for those using lockfiles.

---

WDYT @hoomane @agriyakhetarpal?

**Comment by hoodmane** [comment]
Are there any examples of a shared library used by multiple packages?

**Comment by ryanking13** [comment]
> Are there any examples of a shared library used by multiple packages?

openBLAS is a good example I think, it is used in both `scipy` and `sparseqr` and `sparseqr` depends on scipy. It's also possible to link numpy and openblas, although we don't currently do that. 

libheif is also used in both pillow_heif and pyheif (I used vendor-sharedlib for these packages as I thought they are unlikely used together)

**Comment by rgommers** [comment]
> We are increasingly moving towards out-of-tree builds and encouraging individual package maintainers to build their own packages 

My understanding here is that there's several levels to this effort:

1. Have a CI job in the repo that builds a package with Pyodide. Benefit: avoid regressions for Pyodide and make the life of Pyodide maintainers easier.
2. Upload nightlies somewhere other than PyPI (e.g., https://anaconda.org/scientific-python-nightly-wheels/). Benefit: needed to use these nightly wheels in order to have an up-to-date version of the package itself in interactive docs. Still unsolved, but work is ongoing to do so (CORS headers issue)
3. Upload to PyPI. Seems like a ways off (?).

I had (1) and (2) in mind for the time being, and that has very concrete benefits for Pyodide itself and for package authors. (3) was not on my radar yet, and I see some potential hiccups regarding how Python packaging works that would need resolving (hence my questions above).

**Comment by ryanking13** [comment]
@rgommers Thanks for sharing your valuable experience! As a user of numpy and scipy, I didn't know that manylinux wheels also have some clumsy parts.

> I am probably missing context on the long-term strategy here, but would it be possible for packages uploaded to PyPI to depend on packages providing shared libraries by Pyodide in-tree. Presumably there is a path for this somehow, since the Python runtime needs to come from outside PyPI as well?

I think it is a good idea. I never thought of releasing shared libraries as a wheel, but scipy-openblas32 example gives me some valuable insights of how we can deal with shared libraries.

> because of the constraints on having everything built with the exact same Emscripten version. 

We still have this constraint that the compatibility between Emscripten versions is not guaranteed. We don't have a perfect solution for this yet, but at least we hope to "test" some range of Emscripten versions to test if they are compatible and tag them with the same platform (e.g., pyodide_2024_X).

**Comment by ryanking13** [comment]
> I had (1) and (2) in mind for the time being, and that has very concrete benefits for Pyodide itself and for package authors. (3) was not on my radar yet, and I see some potential hiccups regarding how Python packaging works that would need resolving (hence my questions above).

Yes, I am also experimenting building packages in a [separate repository](https://github.com/ryanking13/pyodide-recipes-mirror) and uploading it to [anaconda package index](https://anaconda.org/pyodide/dashboard). In the long term, the goal is to delegate package management to package maintainers (or recipe maintainers like conda).
